compiler-core/src/type_/expression.rs

```diff
@@ -1046,6 +1046,57 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 }
             }
 
+            ClauseGuard::FieldAccess {
+                location,
+                label,
+                container,
+                ..
+            } => {
+                let typed_container = self.infer_clause_guard(*container.clone())?;
+
+                match typed_container.type_().as_ref() {
+                    Type::App { .. } => match typed_container.clone() {
+                        ClauseGuard::Var { location, name, .. } => {
+                            let container_expr = UntypedExpr::Var { location, name };
+                            match self.infer_field_access(
+                                container_expr,
+                                label.clone(),
+                                location,
+                                FieldAccessUsage::Other,
+                            )? {
+                                TypedExpr::RecordAccess { index, typ, .. } => {
+                                    Ok(ClauseGuard::FieldAccess {
+                                        location,
+                                        label,
+                                        index: Some(index + 1),
+                                        type_: typ,
+                                        container: Box::new(typed_container),
+                                    })
+                                }
+
+                                _ => panic!("Expected RecordAccess"),
+                            }
+                        }
+
+                        ClauseGuard::FieldAccess { .. } => self.infer_clause_guard(*container),
+
+                        _ => Err(Error::NotATuple {
+                            location,
+                            given: typed_container.type_(),
+                        }),
+                    },
+
+                    typ if typ.is_unbound() => Err(Error::NotATupleUnbound {
+                        location: typed_container.location(),
+                    }),
+
+                    _ => Err(Error::NotATuple {
+                        location,
+                        given: typed_container.type_(),
+                    }),
+                }
```

Comment:
removing the `match` with the `Type::App` clause mostly solved it. But there is still the `match` to see whether the container is itself a field access or not which should have a catch all clause to satisfy exhaustiveness but otherwise is not needed.